### DISCUSSION TITLE
Show dynamic number of rows using Variables

### DISCUSSION THREAD

**Discussion by ShermaineMY** [eclipse-birt/birt#1080] [Q&A]
Situation is want to show the correct running number when there is situation that some row will be hidden. 
I got the below suggestion to use variables but not sure how exact to create it into BIRT. 

1. create the variable GridRownumber
2. using dynamic text to put below expression:
   vars["GridRownumber"]
 if(conditiontoshowpreviousGridRow){vars["GridRownumber"]=vars["GridRownumber"]+1;}

Q1: The variables that create, do I need to put in any default value
Q2:  How to put this javascript for this  "conditiontoshowpreviousGridRow". 

Thanks in advance

**Comment by hvbtup** [comment]
I assume that you are using a table.
A1: Initialize the var inside the onCreate script of the table item.
A2: Increment the var inside the visibility expression for the row (the expression can contain statements).
For example, your row visibility expression could look like this:

    var showRow = ...; // should be a boolean
    if (showRow) {
        vars["GridRownumber"] += 1;
    }
    showRow // note there is no semicolon at the end of the last expression

But unless there is a special reason against it, you should use a *filter* on the table item instead.
Then you can use __rownum for a zero-based row numbering.

**Reply by ShermaineMY** [reply]
I am using Grid instead of table. There is fix title for each row, so using the Grid > row.

**Reply by hvbtup** [reply]
OK. The filtering approach doesn't make sense in this case, but the solution using onCreate and visibility expression should work anyway.

**Reply by ShermaineMY** [reply]
Can share more details in exact how to configure?

**Comment by TheLaserNinja** [comment]
Why are you using a grid for something that is based on a dynamic condition? Grids are designed to be static. If you know there are a static number of rows in your grid, its easier to use a scripted data set with a simple counter, and a visibility expression and a running count aggregation to achieve this. 

Either way, you can use this example to illustrate how to use the report context to set a global variable. I've included the table method as described. 

I've also included an example using a Grid. This will set a global variable in the onPrepare script event for the grid. Then, using a Dynamic Text element, it will display the result of the global variable. In the Visibility expression for each row in the grid, I am setting a variable to either true or false (simulating whatever you are using for logic to dynamically hide or show those rows), then checking if that variable is set, if it is false (meaning show the row), getting the global variable, increasing it, then resetting it in the global context. 
[RunningCountWithVisibilityExpressionExample.rptdesign copy.zip](https://github.com/eclipse/birt/files/9721117/RunningCountWithVisibilityExpressionExample.rptdesign.copy.zip)

**Reply by speckyspooky** [reply]
The example of @TheLaserNinja is a good one. 

It is a little bit difficult without a draft of the original report to assist on it (@ShermaineMY).
Another option would be to use "normal" tables with a filtered dataset - filtered means your parameter (if you use parameters) would be filter of your sql-statement and so you would get only the necessary data with correct numbering of __rownum.

**Reply by ShermaineMY** [reply]
Hi All,

I am the newbie as BIRT configuration. I may have difficulty to understand it. I not sure how to open/run @TheLaserNinja sample in BIRT so that I can refer it to configure into my own file.  Able to explain or share more details, especially for the variable part.  I know there is report variable or page variable. I need to create a new for variable then default value.  

 @speckyspooky need to in Grid because content is from different secondary dataset.  

This is the situation:-
![Screenshot (15)](https://user-images.githubusercontent.com/110033368/195059393-89873d14-a843-4143-8af9-c49bfa080431.png)

**Comment by hvbtup** [comment]
This is not related to your original questions, but the Javascript in your visibility expression seems to contain errors.

Are you aware that `&` means a "binary and" (not "logical and")? Probably you should use `&&`.

The operands are comparisons, so they are boolean, and the result is the same in this special case, but this still is a programming error.

**Reply by ShermaineMY** [reply]
Noted on the &&> I will change it. Thanks
